### ISSUE TITLE
cdk deploy: do not use docker for dependency resolution & installation

### ISSUE THREAD

**Issue by CoreOxide** [aws/aws-cdk#25442]
### Describe the feature

Today, CDK deploy uses docker as it's backend to collect and install dependencies. I'm proposing a way to to that without using Docker. It will also allow building for different OS targets and Lambda runtimes.

### Use Case

Not needing to run (and pay for) Docker anymore when building deploying.

### Proposed Solution

This method basically implements pip's dependency resolution. It creates a flat dependency list that can be fed to pip. It's required to have an already resolved dependency list to utilize pip's --_platform_ flag, such that dependencies can be downloaded for any target OS (given they exist for that OS).

See this blog post for more information:
https://medium.com/cyberark-engineering/dont-use-docker-to-package-python-lambdas-on-your-mac-fe64b8f98144

See this repo for working code solution:
https://github.com/CoreOxide/pip_requirements_flattener

### Other Information

This code is used in production-grade deployment pipelines for some time now and has a proved record.

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.69.0

### Environment details (OS name and version, etc.)

MacOS 13.3.1

**Comment by peterwoodworth** [comment]
Is this specifically regarding the lambda python module?

**Comment by CoreOxide** [comment]
> Is this specifically regarding the lambda python module?

Yup. I'm not sure if containers are used for any other languages, though, an if this solution can be expanded to cover them as well.

**Comment by CoreOxide** [comment]
@peterwoodworth So do you think it will be fine to introduce this new dependency resolution & installation system to support only Python Lambdas for the first iteration?
If so, I can create a PR in CDK.

**Comment by peterwoodworth** [comment]
I think we should do an [RFC](https://github.com/aws/aws-cdk-rfcs) for this first, as this seems to be a significant overhaul of the existing implementation

**Comment by CoreOxide** [comment]
Thanks Peter. I'll try to get to it in the coming weeks.

**Comment by kevinhq** [comment]
Any update on this? It would be great if it's implemented. Docker didn't run very well on macOS (at least from my experience, it's slower... than Linux)

**Comment by CoreOxide** [comment]
@kevinhq I want to try another approach, using AWS ClodShell instead. I hope I can get ot it soon.
If it will indeed prove the better method - I will propose it in the RFC instead.

**Comment by CoreOxide** [comment]
@kevinhq , @peterwoodworth 
So I finally gov some time to work on this. Created RFC -> https://github.com/aws/aws-cdk-rfcs/issues/579
Should this issue be closed now?

**Comment by mrpackethead** [comment]
> I think we should do an [RFC](https://github.com/aws/aws-cdk-rfcs) for this first, as this seems to be a significant overhaul of the existing implementation

Except the RFC process is closed, so hows that going to work?

**Comment by evgenyka** [comment]
@CoreOxide - Thank you for submitting your RFC! Please be aware that the RFC process is currently on hold (as we announced in August 2023), and we won't be able to review it immediately. I recommend staying updated with our (near) future announcements regarding the RFC process. We will use this as a tracking issue for your RFC.

**Comment by aemada-aws** [comment]
Looks like a duplicate of https://github.com/aws/aws-cdk/issues/18290
